### PR TITLE
load plugins only when the corresponding instrumentation is loaded

### PR DESCRIPTION
**Pull request by rochdev**
### What does this PR do?

Load plugins only when the corresponding instrumentation is loaded.

### Motivation

Plugins can sometimes do heavy operations when instantiated, and sometimes even just when their file is loaded. By requiring plugins only when they will actually be needed because the module the instrument was loaded, we eliminate their code completely when not needed.

### Additional Notes

This change doesn't play well with `agent.wipe()`, so the call was removed from most places. However, it seems this has exacerbated an existing issue where tests sometimes timeout. I wasn't able to find the cause but I have confirmed it also happens on `master`, just to a much lesser extent. This is something we can investigate later when we rework the tests.

I have also added a new `agent.reload()` function while working on the above known issue, and since right now the only additional change is a few lines in the `amqp10` and `rhea` tests and the function itself, I left it in the PR. Later on we will be able to use this in all plugin tests to avoid constantly destroying and recreating the tracer.

### CODE REVIEW

File: packages/datadog-instrumentations/src/helpers/instrument.js [hunk 1]
```diff
@@ -28,12 +30,22 @@ exports.addHook = function addHook ({ name, versions, file }, hook) {
       return moduleExports
     }
 
-    try {
-      return hook(moduleExports)
-    } catch (e) {
-      log.error(e)
-      return moduleExports
+    if (moduleName === fullFilename) {
```
Comment (juan-fernandez) [inline packages/datadog-instrumentations/src/helpers/instrument.js:33]: wouldn't `moduleName` already be equal to `fullFilename` here? 

Would this work?
```javascript
const version = getVersion(moduleBaseDir)
if (!matchVersion(version, versions)) {
  return moduleExports
}
try {
  loadChannel.publish({ name, version, file })

  return hook(moduleExports)
} catch (e) {
  log.error(e)
  return moduleExports
}
```

**Review by juan-fernandez — COMMENTED** [review]
cool that we're leveraging `diagnostics_channel` for this 👍 . I left a nitpick but overall it looks good to me

File: packages/datadog-instrumentations/src/helpers/instrument.js (hunk 1 above)
Comment (juan-fernandez) [reply packages/datadog-instrumentations/src/helpers/instrument.js:33]: actually the version would also match I think, so we don't even need the `matchVersion` check I think 🤔

File: packages/datadog-instrumentations/src/helpers/instrument.js [hunk 2]
```diff
@@ -28,12 +30,22 @@ exports.addHook = function addHook ({ name, versions, file }, hook) {
...
+      const version = getVersion(moduleBaseDir)
```
Comment (uurien) [inline packages/datadog-instrumentations/src/helpers/instrument.js:34]: `getVersion` is reading the version from the package.json. We are calling to `getVersion` in line 29, that means that we are reading and parsing `package.json` file twice.
I propose something like that:
```suggestion
   let version
    if (moduleName !== fullFilename || !matchVersion((version = getVersion(moduleBaseDir), version), versions)) {
      return moduleExports
    }

    if (moduleName === fullFilename) {
      if (matchVersion(version, versions)) {
```

Comment (rochdev) [reply packages/datadog-instrumentations/src/helpers/instrument.js:34]: I wanted to move the code but instead I ended up just copying it, fixing now.

File: packages/datadog-instrumentations/src/helpers/instrument.js (hunk 1 above)
Comment (rochdev) [reply packages/datadog-instrumentations/src/helpers/instrument.js:33]: I meant to move the code but instead I just copied, fixing now.

**Review by uurien — DISMISSED** [review]

**Review by simon-id — APPROVED** [review]

File: packages/dd-trace/test/plugins/agent.js [hunk 3]
```diff
@@ -75,6 +75,18 @@ module.exports = {
...
+  reload (pluginName, config) {
+    pluginName = [].concat(pluginName)
+    plugins = pluginName
+    config = [].concat(config)
+
+    for (let i = 0, l = pluginName.length; i < l; i++) {
+      tracer.use(pluginName[i], config[i])
+    }
+
+    return Promise.resolve()
```
Comment (bengl) [inline packages/dd-trace/test/plugins/agent.js:87]: Why does this need a promise?

**Review by bengl — CHANGES_REQUESTED** [review]

File: packages/dd-trace/test/plugins/agent.js (hunk 3 above)
Comment (rochdev) [reply packages/dd-trace/test/plugins/agent.js:87]: It doesn't need to be, I originally implemented it directly in `load` then moved it so it's mostly a leftover. I can remove it.

File: packages/datadog-plugin-winston/test/index.spec.js [hunk 4]
```diff
@@ -54,7 +55,7 @@ describe('Plugin', () => {
 
     spy = sinon.spy()
 
-    class Transport extends winston.Transport {}
+    Transport = class extends winston.Transport {}
```
Comment (Qard) [inline packages/datadog-plugin-winston/test/index.spec.js:58]: Why is this change necessary? Doesn't seem to be accessed anywhere else. 🤔

**Review by Qard — COMMENTED** [review]
How does this handle cases where there's a plugin and no instrumentation? Ideally we should move toward having diagnostics_channel events directly in libraries and only needing the subscriber side on our end.

**Comment by rochdev** [comment]
> How does this handle cases where there's a plugin and no instrumentation?

Ideally Node would publish events on every import, but short of that we could use our existing import hook.

File: packages/datadog-plugin-winston/test/index.spec.js (hunk 4 above)
Comment (rochdev) [reply packages/datadog-plugin-winston/test/index.spec.js:58]: The change was needed for another change that was eventually reverted, so it should no longer be needed.

**Review by bengl — APPROVED** [review]

File: packages/datadog-plugin-winston/test/index.spec.js (hunk 4 above)
Comment (rochdev) [reply packages/datadog-plugin-winston/test/index.spec.js]: I reverted the change.

**Review by Qard — APPROVED** [review]
As long as there's no dependency of instrumentation existing for a plugin to be able to load, I'm good with this. We should probably see about augmenting module require/import with diagnostics_channel events sooner rather than later so we have a clean path forward in the near future.
